filewatcher/v2/filewatcher.go

```diff
@@ -0,0 +1,567 @@
+	// counter of limitopen_softlimit_violation_total,
+	// but it does not stop the normal parsing process.
+	//
+	// If the hard limit is violated,
+	// The loading of the file will fail immediately.
+	MaxFileSize int64 `yaml:"maxFileSize"`
+
+	// Optional, the interval to check file changes proactively.
+	//
+	// Default to DefaultPollingInterval.
+	// To disable polling completely, set it to a negative value.
+	//
+	// Without polling filewatcher relies solely on fs events from the parent
+	// directory. This works for most cases but will not work in the cases that
+	// the parent directory will be remount upon change
+	// (for example, k8s ConfigMap).
+	PollingInterval time.Duration `yaml:"pollingInterval"`
+
+	// Optional, the delay between receiving the fs events and actually reading
+	// and parsing the changes.
+	//
+	// It's used to avoid short bursts of fs events (for example, when watching a
+	// directory) causing reading and parsing repetively.
+	//
+	// Defaut to DefaultFSEventsDelay.
+	FSEventsDelay time.Duration `yaml:"fsEventsDelay"`
+
+	// Optional, the interval to keep retrying to open the file when creating a
+	// new FileWatcher, when the file was not initially available.
+	//
+	// Default to DefaultInitialReadInterval.
+	InitialReadInterval time.Duration `yaml:"initialReadInterval"`
+}
+
+type fakeDirectoryReader string
+
+func (fakeDirectoryReader) Read([]byte) (int, error) {
+	return 0, errors.New("filewatcher: you are most likely watching a directory without using DirParser")
+}
+
+func openAndParse[T any](path string, parser Parser[T], limit, hardLimit int64) (data T, mtime time.Time, files []string, _ error) {
+	var zero T
+	stats, err := os.Stat(path)
+	if err != nil {
+		return zero, time.Time{}, nil, fmt.Errorf("filewatcher: i/o error: %w", err)
+	}
+	mtime = stats.ModTime()
+	files = []string{
+		// Note: We need to also watch the parent directory,
+		// because only watching the file won't give us CREATE events,
+		// which will happen with atomic renames.
+		filepath.Dir(path),
+		path,
+	}
+
+	var reader io.Reader
+	if stats.IsDir() {
+		reader = fakeDirectoryReader(path)
+		if err := filepath.Walk(path, func(p string, _ fs.FileInfo, err error) error {
+			if err == nil {
+				files = append(files, p)
+			}
+			return nil
+		}); err != nil {
+			return zero, time.Time{}, nil, fmt.Errorf("filewatcher: i/o error: %w", err)
+		}
+	} else {
+		// file
+		f, err := limitopen.OpenWithLimit(path, limit, hardLimit)
+		if err != nil {
+			return zero, time.Time{}, nil, fmt.Errorf("filewatcher: i/o error: %w", err)
+		}
+		defer f.Close()
+		reader = f
+	}
+
+	d, err := parser(reader)
+	if err != nil {
+		return zero, time.Time{}, nil, fmt.Errorf("filewatcher: parser error: %w", err)
+	}
+	return d, mtime, files, nil
+}
+
+type opts struct {
+	fsEventsDelay       time.Duration
+	pollingInterval     time.Duration
+	initialReadInterval time.Duration
+
+	fileSizeLimit int64
+}
+
+// Option used in New.
+type Option func(*opts)
```

Comment:
Whenever you have an `Option` type, we've come to the conclusion that you should always have a `WithOptions(...Option) Option` helper too.  It makes currying zero or more options into one much easier and avoids the need to append options (and potentially end up with aliasing bugs that can be super hard to track down).

For example, you can do `WithOptions(defaults(), providedOptions...)` to easily combine the default values (expressed as options) with the user-provided overrides.  It also replaces the loop in the constructor, which is now just `var opt Options; WithOptions(defaults(), providedOptions...)(&opt)`